libs/domain/cart/coupon/coupon.component.ts

```diff
@@ -0,0 +1,128 @@
+import { query, state } from 'lit/decorators.js';
+import { repeat } from 'lit/directives/repeat.js';
+import { tap } from 'rxjs';
+import { couponStyles } from './coupon.styles';
+
+@hydrate({ event: ['window:load'] })
+export class CouponComponent extends CartComponentMixin(
+  ContentMixin(LitElement)
+) {
+  static styles = couponStyles;
+
+  @state() hasError = false;
+  @state() errorMessageDescription: '' | string = '';
+
+  @query('input[name=coupon]') coupon?: HTMLInputElement;
+
+  protected cartService = resolve(CartService);
+  protected notificationService = resolve(NotificationService);
+  protected localeService = resolve(LocaleService);
+
+  protected override render(): TemplateResult | void {
+    if (this.$isEmpty()) {
+      return;
+    }
+
+    this.errorMessageDescription =
+      this.coupon?.value === ''
+        ? `${this.i18n('coupon.insert-a-coupon')}`
+        : `${this.i18n('coupon.cart-code-can-not-be-added')}`;
+
+    return html`
+      <oryx-layout>
+        <h3>${this.i18n('coupon.have-a-coupon')}?</h3>
+
+        <section>
+          <oryx-input
+            ?hasError="${this.hasError}"
+            .errorMessage="${this.hasError ? this.errorMessageDescription : ''}"
+          >
+            <input placeholder="Coupon code" name="coupon" />
+            <oryx-icon
+              type=${IconTypes.Close}
+              class="clear-icon"
+              appearance=${ClearIconAppearance.Hover}
+              @click=${(): void => this.clear()}
+            ></oryx-icon>
+          </oryx-input>
+          <oryx-button
+            .type="${ButtonType.Outline}"
+            .color="${ColorType.Neutral}"
+            .size=${ButtonSize.Md}
+            @click=${this.onSubmit}
+          >
+            ${this.i18n('coupon.apply')}
+          </oryx-button>
+        </section>
+      </oryx-layout>
+
+      <ul>
+        ${repeat(
+          this.$coupons(),
+          (coupon) => coupon.code,
+          (coupon) => {
+            let expirationDateTime = '';
+            this.localeService
+              .formatDate(coupon.expirationDateTime)
+              .pipe(
+                tap((date) => {
+                  expirationDateTime = date ?? '';
+                })
+              )
+              .subscribe();
+            return html`
+              <li>
+                <oryx-icon type="done"></oryx-icon>
+                <span class="code">${coupon.code} </span>
+                <span class="name"
+                  >${coupon.displayName}
+                  <span>(valid till ${expirationDateTime})</span>
+                </span>
+              </li>
+            `;
+          }
+        )}
+      </ul>
+    `;
+  }
+
+  protected clear(): void {
+    this.coupon!.value = '';
+  }
+
+  protected onSubmit(): void {
+    this.hasError = false;
+
+    if (this.coupon?.value === '') {
+      this.hasError = true;
+
+      return;
+    }
+
+    this.cartService.addCoupon({ code: this.coupon?.value ?? '' }).subscribe({
```

Comment:
Maybe better to move this to the cartController?

Reply:
I don’t think is necessary to change to controller, it’s a observable that leaves for a short time, we use the same approach to update the 	cart and item entries